Dictionary

DmitryIldyukov opened the pull request:
**Задание: Разработка консольного приложения Dictionary**

**Описание задания**
Вам необходимо разработать консольное приложение на языке C#, которое будет
выполнять роль словаря. Приложение должно предоставлять следующие возможности:
1. Перевод слов с русского на английский и наоборот.
2. Добавление новых слов в словарь.
3. Хранение данных в файле. Файл может быть изначально заполнен или пуст.

**Требования к реализации**

1. Перевод слов: Пользователь вводит слово, и программа выводит его перевод.
Если слово отсутствует в словаре, программа уведомляет пользователя об этом
и предлагает добавить новое слово.
2. Добавление новых слов: Пользователь может добавить новое слово и его перевод в словарь. Программа должна сохранять новые слова в файл для последующего
использования.
3. Хранение данных в файле: Словарь должен загружаться из файла при запуске программы. Все изменения (добавление новых слов) должны сохраняться в файл.

**Рекомендации по работе с файлами на C#**
1. Чтение данных из файла:
Для чтения данных из файла используйте классы StreamReader или методы из
класса File , такие как File.ReadAllLines или File.ReadAllText .
Note: подумайте, хороши ли читать весь текст из файла за раз
Преобразуйте прочитанные данные в структуру, удобную для работы (например,
Dictionary<string, string> ).
2. Запись данных в файл:
Для записи данных в файл используйте классы StreamWriter или методы из
класса File , такие как File.WriteAllLines или File.WriteAllText .
Убедитесь, что все изменения сохраняются в файл при завершении работы
программы.
3. Формат хранения данных:
Вы можете использовать простой текстовый формат, где каждая строка файла
содержит слово и его перевод, разделенные символом-разделителем
(например, запятой или табуляцией).
Пример строки: cat:кот .

DmitryIldyukov committed:
Добавил Dictionary

DmitryIldyukov committed:
Вынес в отдельные файлы

DmitryIldyukov committed:
Исправил форматирование

DmitryIldyukov committed:
Расположил методы в порядке выполнения

Nafanyan commented on Dictionary/Exceptions/RocordNotFoundException.cs:
Имя файла не соответствует классу, который в нём находится

Nafanyan commented on Dictionary/DictionaryService.cs:
```diff
@@ -0,0 +1,92 @@
+    /// </summary>
+    /// <param name="word">Слово.</param>
+    /// <param name="wordTranslate">Перевод слова.</param>
+    /// <exception cref="RecordAlreadyExistException">Выбрасывается, если слово или его перевод уже существуют в словаре.</exception>
```
Я как пользователь не знал об этой особенности и думал, что смогу изменить перевод, но это не так. Лучше уведомить пользователя о таких неочевидных вещах)
![image](https://github.com/DmitryIldyukov/TLPractice/assets/90004716/6dcc201c-1f9c-4262-9bc6-d5b53fc6c069)

DmitryIldyukov replied:
Теперь можно изменять перевод)

Nafanyan commented on Dictionary/DictionaryService.cs:
```diff
@@ -0,0 +1,92 @@
+        }
+    }
+
+    private bool RecordCanBeAdded( string word, string translate ) =>
```
получается очень длинная строка, в таких случаях лучше делать 
```c#
    private bool RecordCanBeAdded( string word, string translate )
    {
        return !_dictionary.Any( record => record.Key == word 
            || record.Value == word 
            || record.Key == translate 
            || record.Value == translate );
    }
```

Nafanyan replied:
Может CanAddRecord ?

Nafanyan commented on Dictionary/DictionaryService.cs:
```diff
@@ -0,0 +1,92 @@
+            return translate;
+        }
+
+        var key = _dictionary.FirstOrDefault( rec => rec.Value == word ).Key;
```
var лучше использовать когда из правой части очевидно какой тип данных ты кладешь в переменную, например:
var key = "key"
var person = new Person()
var num = 12

Nafanyan replied:
а можешь объяснить зачем тебе эта логика? ты как будто второй раз ищешь перевод. При каком кейсе ты получишь key != null?

DmitryIldyukov replied:
Я немного намудрил с логикой, наверное, лучше было бы обрабатывать отдельные usecase-ы (перевести с русского на английски, перевести с английского на русский и т.п.) 
Сейчас сделано так, что пользователь может ввести слово на любом языке, поэтому нужно проверить и key, и value у словаря)

Nafanyan commented on Dictionary/DictionaryService.cs:
```diff
@@ -0,0 +1,92 @@
+            {
+                words[ 0 ] = words[ 0 ].ToLower().Trim();
+                words[ 1 ] = words[ 1 ].ToLower().Trim();
+                if ( RecordCanBeAdded( words[ 0 ], words[ 1 ] ) )
```
а зачем здесь нужно это условие, страхует от возможных дубликатов в файле?

DmitryIldyukov replied:
Да)

Nafanyan commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,130 @@
+        Console.WriteLine( "Спасибо, что воспользовались приложением! До свидания!" );
+    }
+
+    static string GetNonEmptyStringFromConsole( string message, string errorMessage = "Неверный ввод. Введенная строка не может быть пустой. Попробуйте снова." )
```
это GPTшка подсказала такую функцию?) если нет то прикольно, удобно

DmitryIldyukov replied:
Название да, использовал еще в прошлом PR)

Nafanyan commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,130 @@
+
+    public static void AddTranslate( string word = null )
+    {
+        if ( word == null )
```
Может лучше так? 
``` c#
string.IsNullOrEmpty( word )
```

Nafanyan commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    public static void GetAllTranslates()
```
неплохо, доп функция, мелочь, а приятно)

Nafanyan commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    public static void AddTranslate( string word = null )
```
Translate - глагол
Translation - существительно

Название как будто немного некорректное, к остальным названиям тот же самый вопрос)

название переменных должны быть существительными (переведенное слово, слово, команда)
название методов глаголами (перевести слово, добавить перевод, очистить словарь)

Nafanyan commented on Dictionary/DictionaryService.cs:
```diff
@@ -0,0 +1,92 @@
+
+    public DictionaryService()
+    {
+        GetWordsFromFile();
```
ты назвал GetWordsFromFile, но никакие слова ты не возвращаешь)

Подумай над названием)

Nafanyan commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,130 @@
+            switch ( Console.ReadLine() )
+            {
+                case "1":
+                    GetTranslate();
```
GetTranslate, но ты ничего не возвращаешь. Когда используется Get подразумевается что функция возвращает какой-то результат, это вводит в заблуждение)

Nafanyan commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,130 @@
+                    AddTranslate();
+                    break;
+                case "3":
+                    GetAllTranslates();
```
Аналогично с Get'ом и здесь, функция void

Nafanyan commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,130 @@
+    {
+        string input;
+
+        while ( true )
```
можешь через do while цикл, тогда не надо будет создавать бесконечный и выходить с помощью break

DmitryIldyukov replied:
Менял в прошлом PR с do-while на while, т.к. при использовании do-while происходила двойная проверка:

```
static string GetNonEmptyStringFromConsole( string message, string errorMessage )
    {
        string input;
        do
        {
            Console.Write( message );
            input = Console.ReadLine();

            if ( string.IsNullOrWhiteSpace( input ) )
                Console.WriteLine( errorMessage );

        } while ( string.IsNullOrWhiteSpace( input ) );

        return input;
    }
```

Как лучше поступить?

Nafanyan replied:
ладно, давай оставим так, это не критично)

Nafanyan reviewed (changes requested):
Заморочился, вынес логику по работе со словарем в отдельный сервис, хорошо придумано)

Однако есть некоторые моменты на которые стоит обратить внимание, описал в комментах, если будут вопросы пиши)

DmitryIldyukov committed:
Исправил замечания

1) Добавил 2 новых действия: изменение перевода и удаление перевода
2) Выделил интерфейс IDictionaryService
3) Переименовал методы

Nafanyan reviewed (approved):
Всё отлично, можешь squash'ить и мержить)
